File: crates/lvz-protocol/src/event.rs
//! The normalised event stream — the semantic layer every provider maps onto.
//!
//! Each provider adapter is the *only* place that translates its wire format into these
//! variants (§5.2): Anthropic SSE `content_block_delta`, xAI gRPC v6 outputs,
//! and the OpenAI-compat fallback all converge here. Nothing downstream of a provider sees
//! a wire protocol.

use serde::{Deserialize, Serialize};

/// A single normalised event in a streamed model turn.
///
/// Adjacently tagged (`{"kind": …, "data": …}`) so every variant — including the newtype
/// variants wrapping a primitive ([`TextDelta`](Event::TextDelta), [`Done`](Event::Done)) —
/// round-trips through JSON. This is the on-the-wire shape gateways stream to their channels.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", content = "data", rename_all = "snake_case")]
pub enum Event {
    /// Incremental assistant text.
    TextDelta(String),
    /// Incremental extended-thinking text (Anthropic). Providers without thinking never emit this.
    Thinking(String),
    /// A tool call has begun; `id` correlates the following deltas and end.
    ToolUseStart { id: String, name: String },
    /// Incremental tool-argument JSON for the call identified by `id`.
    ToolUseDelta { id: String, json: String },
    /// The tool call identified by `id` is complete; its argument JSON is now whole.
    ToolUseEnd { id: String },
    /// A **provider-executed** (server-side) tool was invoked by the model — e.g. web search or
    /// code execution. Informational: the agent does not execute these (the provider does).
    ServerToolUse { id: String, name: String },
    /// The result of a provider-executed tool, as a serialized JSON string (search hits, code
    /// stdout/stderr, fetched page, …), correlated by `id`.
    ServerToolResult { id: String, content: String },
    /// A citation the model attached to the preceding span of output (Anthropic document /
    /// web-search citations, emitted as the text is produced). Informational — the agent
    /// forwards it; gateways/CLIs surface it.
    Citation {
        /// The exact text from the source that supports the cited output span.
        cited_text: String,
        /// Human-readable source reference (document title, URL, or `document N`).
        source: String,
    },
    /// Token accounting, including cache hits. May arrive mid-stream and/or at the end.
    Usage(Usage),
    /// Terminal event: the turn finished for the given reason.
    Done(StopReason),
}

/// Token accounting for a turn. Cache fields are zero on providers without prompt caching.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct Usage {
    pub input_tokens: u64,
    pub output_tokens: u64,
    /// Tokens written to the cache this turn (Anthropic `cache_creation_input_tokens`).
    pub cache_creation_tokens: u64,
    /// Tokens served from the cache this turn (Anthropic `cache_read_input_tokens`).
    pub cache_read_tokens: u64,
}

/// Per-token-class cost multipliers, expressed relative to a fresh input token (= 1.0).
///
/// Turns a raw [`Usage`] into a single **cost-weighted** objective ([`Usage::cost`]) so the
/// agent's budget and the ATO tuner optimise *actual spend*, not a flat token count. This is
/// what makes prompt caching — the project's biggest cost lever (§6) — visible to
/// optimisation: a fresh input token, an output token, a cache *write*, and a cache *read* all
/// cost different amounts, and a flat sum (or one that ignores cache classes outright) hides
/// that. The defaults match Anthropic/xAI list-price ratios, which are also a sensible
/// cross-provider baseline.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct CostWeights {
    pub input: f64,
    pub output: f64,
    pub cache_creation: f64,
    pub cache_read: f64,
}

impl Default for CostWeights {
    /// Anthropic/xAI ratios: output ≈ 5× input, a cache write ≈ 1.25× input, a cache read ≈
    /// 0.1× input. A reasonable default for any caching provider.
    fn default() -> Self {
        Self {
            input: 1.0,
            output: 5.0,
            cache_creation: 1.25,
            cache_read: 0.1,
        }
    }
}

impl CostWeights {
    /// Anthropic list-price ratios (Sonnet/Haiku/Opus all share 1 / 5 / 1.25 / 0.1).
    pub fn anthropic() -> Self {
        Self::default()
    }

    /// xAI (grok) ratios — same input/output 1:5 shape; xAI caches automatically and bills the
    /// cached read cheaply, with no separate request-side cache-write line.
    pub fn xai() -> Self {
        Self::default()
    }

    /// Gemini ratios: a wider output:input spread and a pricier cache read than Anthropic.
    pub fn google() -> Self {
        Self {
            input: 1.0,
            output: 8.0,
            cache_creation: 1.0,
            cache_read: 0.25,
        }
    }

    /// Flat weights — every token class counts the same. Restores the legacy raw-token
    /// objective (and is what providers without real caching effectively use).
    pub fn flat() -> Self {
        Self {
            input: 1.0,
            output: 1.0,
            cache_creation: 1.0,
            cache_read: 1.0,
        }
    }
}

impl Usage {
    /// Raw billable input + output tokens. A flat count for display/diagnostics; the budget and
    /// the ATO objective use the cost-weighted [`cost`](Self::cost) instead. Note this does
    /// **not** include the cache classes (which a caching provider reports separately).
    pub fn total(&self) -> u64 {
        self.input_tokens + self.output_tokens
    }

    /// Cost-weighted total in *fresh-input-token-equivalent* units (rounded). This — not the
    /// flat [`total`](Self::total) — is the objective the agent budget enforces and the ATO
    /// tuner minimises, so caching (cheap reads, pricier writes) and output cost all register.
    pub fn cost(&self, w: &CostWeights) -> u64 {
        (self.input_tokens as f64 * w.input
            + self.output_tokens as f64 * w.output
            + self.cache_creation_tokens as f64 * w.cache_creation
            + self.cache_read_tokens as f64 * w.cache_read)
            .round() as u64
    }

    /// Add another turn's usage into this one — the running total across round-trips (§6.4).
    pub fn accumulate(&mut self, other: &Usage) {
        self.input_tokens += other.input_tokens;
        self.output_tokens += other.output_tokens;
        self.cache_creation_tokens += other.cache_creation_tokens;
        self.cache_read_tokens += other.cache_read_tokens;
    }

    /// Fraction of input tokens served from cache, in `[0, 1]`. Zero when no input tokens.
    pub fn cache_hit_rate(&self) -> f32 {
        let billed_input = self.input_tokens + self.cache_read_tokens;
        if billed_input == 0 {
            0.0
        } else {
            self.cache_read_tokens as f32 / billed_input as f32
        }
    }
}

/// Why a turn stopped.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StopReason {
    /// Model finished its turn naturally.
    EndTurn,
    /// Hit the `max_tokens` ceiling.
    MaxTokens,
    /// Stopped to hand control back for tool execution.
    ToolUse,
    /// Matched a caller-supplied stop sequence.
    StopSequence,
    /// The model declined the request for safety reasons (Anthropic `refusal`). Terminal.
    Refusal,
    /// The provider paused a server-side tool loop; resend the turn to continue (Anthropic
    /// `pause_turn`). Not terminal — the caller re-submits to let the model carry on.
    PauseTurn,
    /// Anything provider-specific not captured above.
    Other(String),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn every_event_variant_roundtrips_through_json() {
        let events = vec![
            Event::TextDelta("hello".into()),
            Event::Thinking("hmm".into()),
            Event::ToolUseStart {
                id: "call_1".into(),
                name: "shell".into(),
            },
            Event::ToolUseDelta {
                id: "call_1".into(),
                json: "{\"cmd\":\"ls\"}".into(),
            },
            Event::ToolUseEnd {
                id: "call_1".into(),
            },
            Event::Citation {
                cited_text: "the sky is blue".into(),
                source: "doc.txt".into(),
            },
            Event::Usage(Usage {
                input_tokens: 1,
                output_tokens: 2,
                cache_creation_tokens: 3,
                cache_read_tokens: 4,
            }),
            Event::Done(StopReason::EndTurn),
            Event::Done(StopReason::Other("time_limit".into())),
        ];
        for ev in events {
            let json = serde_json::to_string(&ev).expect("serialize");
            let back: Event = serde_json::from_str(&json).expect("deserialize");
            assert_eq!(ev, back, "roundtrip mismatch for {json}");
        }
    }

    #[test]
    fn cost_weights_make_caching_visible() {
        let w = CostWeights::default();
        // 100 fresh input + 20 output + 1000 cache write + 5000 cache read.
        let usage = Usage {
            input_tokens: 100,
            output_tokens: 20,
            cache_creation_tokens: 1000,
            cache_read_tokens: 5000,
        };
        // Flat total ignores both cache classes entirely.
        assert_eq!(usage.total(), 120);
        // Cost: 100*1 + 20*5 + 1000*1.25 + 5000*0.1 = 100 + 100 + 1250 + 500 = 1950.
        assert_eq!(usage.cost(&w), 1950);
        // Flat weights reduce cost() to summing every class equally.
        assert_eq!(usage.cost(&CostWeights::flat()), 100 + 20 + 1000 + 5000);
        // A cache-churning run (re-writing instead of reading) is now strictly more expensive,
        // which the flat total() could never show.
        let churn = Usage {
            cache_creation_tokens: 6000,
            cache_read_tokens: 0,
            ..usage
        };
        assert!(churn.cost(&w) > usage.cost(&w));
    }

    #[test]
    fn text_delta_has_the_adjacently_tagged_wire_shape() {
        let json = serde_json::to_value(Event::TextDelta("hi".into())).unwrap();
        assert_eq!(json["kind"], "text_delta");
        assert_eq!(json["data"], "hi");
    }
}

File: src/Lavoisier/Protocol/Event.hs
-- | The normalised event stream — the semantic layer every provider maps onto.
--
-- Each provider adapter is the /only/ place that translates its wire format into these
-- constructors; nothing downstream of a provider sees a wire protocol. Ported from the Rust
-- @lvz-protocol@ @event.rs@; the JSON encodings below reproduce that crate's serde shapes exactly
-- (this is the on-the-wire form the gateways stream), so a Rust client and a Haskell gateway agree.
module Lavoisier.Protocol.Event
  ( Event (..),
    Usage (..),
    StopReason (..),
    CostWeights (..),
    emptyUsage,
    usageTotal,
    usageCost,
    accumulateUsage,
    cacheHitRate,
    defaultCostWeights,
    anthropicWeights,
    xaiWeights,
    googleWeights,
    flatWeights,
  )
where

import Data.Aeson
import Data.Aeson.Types (Parser, typeMismatch)
import Data.Text (Text)
import Data.Word (Word64)

-- | A single normalised event in a streamed model turn.
--
-- Adjacently tagged on the wire (@{"kind": …, "data": …}@, snake_case), matching Rust's
-- @#[serde(tag = "kind", content = "data", rename_all = "snake_case")]@.
data Event
  = -- | Incremental assistant text.
    TextDelta Text
  | -- | Incremental extended-thinking text (Anthropic).
    Thinking Text
  | -- | A tool call has begun: @id@ correlates the following deltas and end, then @name@.
    ToolUseStart Text Text
  | -- | Incremental tool-argument JSON (@id@, then a @json@ fragment) for the call.
    ToolUseDelta Text Text
  | -- | The tool call identified by @id@ is complete.
    ToolUseEnd Text
  | -- | A provider-executed (server-side) tool was invoked (@id@, @name@). Informational.
    ServerToolUse Text Text
  | -- | The result of a provider-executed tool (@id@, then @content@ as a JSON string).
    ServerToolResult Text Text
  | -- | A citation the model attached to the preceding output span (@cited_text@, @source@).
    Citation Text Text
  | -- | Token accounting, including cache hits. May arrive mid-stream and/or at the end.
    Usage Usage
  | -- | A human-readable progress notice about the turn (injected by the agent, not the provider).
    Notice Text
  | -- | Terminal event: the turn finished for the given reason.
    Done StopReason
  deriving stock (Eq, Show)

-- | Token accounting for a turn. Cache fields are zero on providers without prompt caching.
data Usage = MkUsage
  { inputTokens :: Word64,
    outputTokens :: Word64,
    cacheCreationTokens :: Word64,
    cacheReadTokens :: Word64
  }
  deriving stock (Eq, Show)

-- | The all-zero usage (Rust @Usage::default()@).
emptyUsage :: Usage
emptyUsage = MkUsage 0 0 0 0

-- | Raw billable input + output tokens (excludes the cache classes) — for display/diagnostics.
usageTotal :: Usage -> Word64
usageTotal u = inputTokens u + outputTokens u

-- | Cost-weighted total in fresh-input-token-equivalent units (rounded) — the budget/ATO objective.
usageCost :: Usage -> CostWeights -> Word64
usageCost u w =
  round $
    fromIntegral (inputTokens u) * cwInput w
      + fromIntegral (outputTokens u) * cwOutput w
      + fromIntegral (cacheCreationTokens u) * cwCacheCreation w
      + fromIntegral (cacheReadTokens u) * cwCacheRead w

-- | Add another turn's usage into this one — the running total across round-trips.
accumulateUsage :: Usage -> Usage -> Usage
accumulateUsage a b =
  MkUsage
    (inputTokens a + inputTokens b)
    (outputTokens a + outputTokens b)
    (cacheCreationTokens a + cacheCreationTokens b)
    (cacheReadTokens a + cacheReadTokens b)

-- | Fraction of input tokens served from cache, in @[0, 1]@. Zero when no input tokens.
cacheHitRate :: Usage -> Double
cacheHitRate u =
  let billed = inputTokens u + cacheReadTokens u
   in if billed == 0 then 0 else fromIntegral (cacheReadTokens u) / fromIntegral billed

-- | Per-token-class cost multipliers, relative to a fresh input token (= 1.0).
data CostWeights = CostWeights
  { cwInput :: Double,
    cwOutput :: Double,
    cwCacheCreation :: Double,
    cwCacheRead :: Double
  }
  deriving stock (Eq, Show)

-- | Anthropic/xAI ratios: output ≈ 5×, cache write ≈ 1.25×, cache read ≈ 0.1×.
defaultCostWeights :: CostWeights
defaultCostWeights = CostWeights 1.0 5.0 1.25 0.1

anthropicWeights :: CostWeights
anthropicWeights = defaultCostWeights

xaiWeights :: CostWeights
xaiWeights = defaultCostWeights

-- | Gemini ratios: wider output:input spread and a pricier cache read than Anthropic.
googleWeights :: CostWeights
googleWeights = CostWeights 1.0 8.0 1.0 0.25

-- | Flat weights — every token class counts the same (the legacy raw-token objective).
flatWeights :: CostWeights
flatWeights = CostWeights 1.0 1.0 1.0 1.0

-- | Why a turn stopped. Externally tagged snake_case on the wire, matching Rust.
data StopReason
  = EndTurn
  | MaxTokens
  | ToolUse
  | StopSequence
  | Refusal
  | PauseTurn
  | -- | Anything provider-specific not captured above.
    Other Text
  deriving stock (Eq, Show)

-- --- JSON: reproduce the Rust serde shapes exactly -------------------------------------------------

instance ToJSON Usage where
  toJSON u =
    object
      [ "input_tokens" .= inputTokens u,
        "output_tokens" .= outputTokens u,
        "cache_creation_tokens" .= cacheCreationTokens u,
        "cache_read_tokens" .= cacheReadTokens u
      ]

instance FromJSON Usage where
  parseJSON = withObject "Usage" $ \o ->
    MkUsage
      <$> o .: "input_tokens"
      <*> o .: "output_tokens"
      <*> o .: "cache_creation_tokens"
      <*> o .: "cache_read_tokens"

instance ToJSON StopReason where
  toJSON = \case
    EndTurn -> "end_turn"
    MaxTokens -> "max_tokens"
    ToolUse -> "tool_use"
    StopSequence -> "stop_sequence"
    Refusal -> "refusal"
    PauseTurn -> "pause_turn"
    Other t -> object ["other" .= t]

instance FromJSON StopReason where
  parseJSON v = case v of
    String s -> case s of
      "end_turn" -> pure EndTurn
      "max_tokens" -> pure MaxTokens
      "tool_use" -> pure ToolUse
      "stop_sequence" -> pure StopSequence
      "refusal" -> pure Refusal
      "pause_turn" -> pure PauseTurn
      _ -> fail ("unknown stop reason: " <> show s)
    Object o -> Other <$> o .: "other"
    _ -> typeMismatch "StopReason" v

instance ToJSON Event where
  toJSON = \case
    TextDelta t -> tagged "text_delta" (toJSON t)
    Thinking t -> tagged "thinking" (toJSON t)
    ToolUseStart i n -> tagged "tool_use_start" (object ["id" .= i, "name" .= n])
    ToolUseDelta i j -> tagged "tool_use_delta" (object ["id" .= i, "json" .= j])
    ToolUseEnd i -> tagged "tool_use_end" (object ["id" .= i])
    ServerToolUse i n -> tagged "server_tool_use" (object ["id" .= i, "name" .= n])
    ServerToolResult i c -> tagged "server_tool_result" (object ["id" .= i, "content" .= c])
    Citation ct s -> tagged "citation" (object ["cited_text" .= ct, "source" .= s])
    Usage u -> tagged "usage" (toJSON u)
    Notice t -> tagged "notice" (toJSON t)
    Done sr -> tagged "done" (toJSON sr)
    where
      tagged :: Text -> Value -> Value
      tagged k d = object ["kind" .= k, "data" .= d]

instance FromJSON Event where
  parseJSON = withObject "Event" $ \o -> do
    kind <- o .: "kind" :: Parser Text
    let dat :: (FromJSON a) => Parser a
        dat = o .: "data"
        obj :: Parser Object
        obj = o .: "data"
    case kind of
      "text_delta" -> TextDelta <$> dat
      "thinking" -> Thinking <$> dat
      "tool_use_start" -> obj >>= \d -> ToolUseStart <$> d .: "id" <*> d .: "name"
      "tool_use_delta" -> obj >>= \d -> ToolUseDelta <$> d .: "id" <*> d .: "json"
      "tool_use_end" -> obj >>= \d -> ToolUseEnd <$> d .: "id"
      "server_tool_use" -> obj >>= \d -> ServerToolUse <$> d .: "id" <*> d .: "name"
      "server_tool_result" -> obj >>= \d -> ServerToolResult <$> d .: "id" <*> d .: "content"
      "citation" -> obj >>= \d -> Citation <$> d .: "cited_text" <*> d .: "source"
      "usage" -> Usage <$> dat
      "notice" -> Notice <$> dat
      "done" -> Done <$> dat
      _ -> fail ("unknown event kind: " <> show kind)
